# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [21]:
# importando as bibliotecas
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt

# from sklearn.model_selection import train_test_split
# from sklearn import metrics
# from sklearn.ensemble import RandomForestClassifier

# from scipy.stats import ks_2samp
import statsmodels.formula.api as smf
import statsmodels.api as sm
import patsy

In [22]:
# abrindo o df
df = pd.read_csv('previsao_de_renda.csv')

In [23]:
# printando as variáveis e seus tipos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             15000 non-null  int64  
 1   data_ref               15000 non-null  object 
 2   id_cliente             15000 non-null  int64  
 3   sexo                   15000 non-null  object 
 4   posse_de_veiculo       15000 non-null  bool   
 5   posse_de_imovel        15000 non-null  bool   
 6   qtd_filhos             15000 non-null  int64  
 7   tipo_renda             15000 non-null  object 
 8   educacao               15000 non-null  object 
 9   estado_civil           15000 non-null  object 
 10  tipo_residencia        15000 non-null  object 
 11  idade                  15000 non-null  int64  
 12  tempo_emprego          12427 non-null  float64
 13  qt_pessoas_residencia  15000 non-null  float64
 14  renda                  15000 non-null  float64
dtypes:

In [24]:
# excluindo linhas duplicadas
df.drop_duplicates()

Unnamed: 0    data_ref  id_cliente sexo  posse_de_veiculo  \
0               0  2015-01-01       15056    F             False   
1               1  2015-01-01        9968    M              True   
2               2  2015-01-01        4312    F              True   
3               3  2015-01-01       10639    F             False   
4               4  2015-01-01        7064    M              True   
...           ...         ...         ...  ...               ...   
14995       14995  2016-03-01       16006    F             False   
14996       14996  2016-03-01        3722    F             False   
14997       14997  2016-03-01        6194    F              True   
14998       14998  2016-03-01        4922    M              True   
14999       14999  2016-03-01       14651    F             False   

       posse_de_imovel  qtd_filhos        tipo_renda           educacao  \
0                 True           0        Empresário         Secundário   
1                 True           0       Assalariado  Superior completo   
2                 True           0        Empresário  Superior completo   
3                 True           1  Servidor público  Superior completo   
4                False           0       Assalariado         Secundário   
...                ...         ...               ...                ...   
14995             True           0        Empresário         Secundário   
14996             True           0       Pensionista  Superior completo   
14997             True           0       Assalariado  Superior completo   
14998            False           0        Empresário  Superior completo   
14999             True           2  Servidor público         Secundário   

      estado_civil tipo_residencia  idade  tempo_emprego  \
0         Solteiro            Casa     26       6.602740   
1           Casado            Casa     28       7.183562   
2           Casado            Casa     35       0.838356   
3           Casado            Casa     30       4.846575   
4         Solteiro   Governamental     33       4.293151   
...            ...             ...    ...            ...   
14995     Solteiro            Casa     48      13.887671   
14996     Solteiro            Casa     57            NaN   
14997       Casado            Casa     45       7.832877   
14998       Casado            Casa     36       4.298630   
14999        União            Casa     34       7.219178   

       qt_pessoas_residencia     renda  
0                        1.0   8060.34  
1                        2.0   1852.15  
2                        2.0   2253.89  
3                        3.0   6600.77  
4                        1.0   6475.97  
...                      ...       ...  
14995                    1.0   7990.58  
14996                    1.0  10093.45  
14997                    2.0    604.82  
14998                    2.0   3352.27  
14999                    4.0   1160.99  

[15000 rows x 15 columns]

In [25]:
# excluindo valores nulos
df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12427 entries, 0 to 14999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             12427 non-null  int64  
 1   data_ref               12427 non-null  object 
 2   id_cliente             12427 non-null  int64  
 3   sexo                   12427 non-null  object 
 4   posse_de_veiculo       12427 non-null  bool   
 5   posse_de_imovel        12427 non-null  bool   
 6   qtd_filhos             12427 non-null  int64  
 7   tipo_renda             12427 non-null  object 
 8   educacao               12427 non-null  object 
 9   estado_civil           12427 non-null  object 
 10  tipo_residencia        12427 non-null  object 
 11  idade                  12427 non-null  int64  
 12  tempo_emprego          12427 non-null  float64
 13  qt_pessoas_residencia  12427 non-null  float64
 14  renda                  12427 non-null  float64
dtypes: bool

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.  


2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.  


3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor. 
    

1. 

In [29]:
# criando a fórmula para o modelo de regressão incluindo todas as variáveis explicativas
y,X = patsy.dmatrices('np.log(renda) ~ C(sexo) + C(posse_de_veiculo) + C(posse_de_imovel) + qtd_filhos + C(tipo_renda) + C(educacao) + C(estado_civil) + C(tipo_residencia) + idade + tempo_emprego + qt_pessoas_residencia',df)

In [30]:
# ajustando e exibindo o modelo
modelo = sm.OLS(y, X).fit()

modelo.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.357
Model:                            OLS   Adj. R-squared:                  0.356
Method:                 Least Squares   F-statistic:                     287.5
Date:                Mon, 21 Oct 2024   Prob (F-statistic):               0.00
Time:                        19:39:59   Log-Likelihood:                -13568.
No. Observations:               12427   AIC:                         2.719e+04
Df Residuals:                   12402   BIC:                         2.737e+04
Df Model:                          24                                         
Covariance Type:            nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                               6.5847      0.235     28.006      0.000       6.124       7.046
C(sexo)[T.M]                            0.7874      0.015     53.723      0.000       0.759       0.816
C(posse_de_veiculo)[T.True]             0.0441      0.014      3.119      0.002       0.016       0.072
C(posse_de_imovel)[T.True]              0.0829      0.014      5.926      0.000       0.055       0.110
C(tipo_renda)[T.Bolsista]               0.2209      0.241      0.916      0.360      -0.252       0.694
C(tipo_renda)[T.Empresário]             0.1551      0.015     10.387      0.000       0.126       0.184
C(tipo_renda)[T.Pensionista]           -0.3087      0.241     -1.280      0.201      -0.782       0.164
C(tipo_renda)[T.Servidor público]       0.0576      0.022      2.591      0.010       0.014       0.101
C(educacao)[T.Pós graduação]            0.1071      0.159      0.673      0.501      -0.205       0.419
C(educacao)[T.Secundário]              -0.0141      0.072     -0.196      0.844      -0.155       0.127
C(educacao)[T.Superior completo]        0.0938      0.072      1.298      0.194      -0.048       0.235
C(educacao)[T.Superior incompleto]     -0.0432      0.078     -0.555      0.579      -0.196       0.109
C(estado_civil)[T.Separado]             0.3265      0.111      2.935      0.003       0.108       0.545
C(estado_civil)[T.Solteiro]             0.2672      0.109      2.453      0.014       0.054       0.481
C(estado_civil)[T.União]               -0.0343      0.025     -1.361      0.174      -0.084       0.015
C(estado_civil)[T.Viúvo]                0.3689      0.116      3.191      0.001       0.142       0.596
C(tipo_residencia)[T.Casa]             -0.0442      0.054     -0.815      0.415      -0.150       0.062
C(tipo_residencia)[T.Com os pais]      -0.0258      0.060     -0.427      0.670      -0.144       0.093
C(tipo_residencia)[T.Comunitário]      -0.1196      0.105     -1.135      0.256      -0.326       0.087
C(tipo_residencia)[T.Estúdio]           0.0665      0.099      0.670      0.503      -0.128       0.261
C(tipo_residencia)[T.Governamental]    -0.0571      0.066     -0.865      0.387      -0.186       0.072
qtd_filhos                             -0.2658      0.108     -2.463      0.014      -0.477      -0.054
idade                                   0.0053      0.001      6.624      0.000       0.004       0.007
tempo_emprego                           0.0617      0.001     59.398      0.000       0.060       0.064
qt_pessoas_residencia                   0.2926      0.108      2.716      0.007       0.081       0.504
==============================================================================
Omnibus:                        0.858   Durbin-Watson:                   2.023
Prob(Omnibus)

In [34]:
y, X = patsy.dmatrices('np.log(renda) ~ C(sexo) + C(posse_de_veiculo) + C(posse_de_imovel) + qtd_filhos  + idade + tempo_emprego', df)

# ajustando e exibindo o modelo
modelo = sm.OLS(y, X).fit()

modelo.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.347
Model:                            OLS   Adj. R-squared:                  0.346
Method:                 Least Squares   F-statistic:                     1098.
Date:                Mon, 21 Oct 2024   Prob (F-statistic):               0.00
Time:                        19:40:50   Log-Likelihood:                -13673.
No. Observations:               12427   AIC:                         2.736e+04
Df Residuals:                   12420   BIC:                         2.741e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       7.2223      0.034    213.646      0.000       7.156       7.289
C(sexo)[T.M]                    0.7688      0.015     52.768      0.000       0.740       0.797
C(posse_de_veiculo)[T.True]     0.0561      0.014      3.986      0.000       0.029       0.084
C(posse_de_imovel)[T.True]      0.0866      0.014      6.278      0.000       0.060       0.114
qtd_filhos                      0.0239      0.009      2.767      0.006       0.007       0.041
idade                           0.0049      0.001      6.399      0.000       0.003       0.006
tempo_emprego                   0.0610      0.001     59.084      0.000       0.059       0.063
==============================================================================
Omnibus:                        1.243   Durbin-Watson:                   2.026
Prob(Omnibus):                  0.537   Jarque-Bera (JB):                1.216
Skew:                           0.022   Prob(JB):                        0.545
Kurtosis:                       3.020   Cond. No.                         223.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

O modelo final simplificado tem menos variáveis, todas significantes, com uma pequena redução no R-squared (0.347 vs. 0.357), o que sugere que a simplicidade foi alcançada sem perda substancial de explicação.